### ISSUE TITLE
WARN Caps are averagely centred — v metrics check

### ISSUE THREAD

**Issue by RosaWagner** [googlefonts/fontbakery#4139]
`capsHeight - hhea/typoAscender ~= - hhea/typoDescender` (this is only recommended for fonts with Latin, Greek or Cyrillic as "main" script, I would go for allowing an error margin of 5% of UPM)

We also had the experience with Braah One: https://github.com/google/fonts/issues/6227
So also in case of non-caps fonts, we could have a similar rule.

**Comment by eliheuer** [comment]
This would be a great check, there is a twitter thread going over the rationale in detail here: https://twitter.com/romanshamin_en/status/1562801657691672576

**Comment by eliheuer** [comment]
I made a pull request for a new check covering this issue here: https://github.com/googlefonts/fontbakery/pull/4226

**Comment by tphinney** [comment]
tl;dr: you might want to do more testing and investigation before implementing that check?

I just want to point out that there is a big assumption behind this, which I am very skeptical of: that all relevant environments people care about, do vertical centering of text using the same algorithm.

I am especially a bit suspicious of the implied claim that em-square as a distance from the origin is one of the metrics in question. I would also be curious as to WHICH descender metric is relevant, or whether that might vary as well.

In the case of Google’s own Flutter web framework, it appears to rely on sTypo metrics including sTypoAscender, and also possibly on… the baseline (?!). But I did not get enough feedback to be sure what was going on. See
https://github.com/google/material-design-icons/issues/1500

**Comment by eliheuer** [comment]
@tphinney Thanks for the feedback! I made a codepen here to try showing why this is a real issue: https://codepen.io/eliheuer/pen/KKrOamg

This kind of vertical centering works on the Web (Chrome) which I would argue is the most relevant environment. And to me it also just intuitively makes sense that you would want to center things this way. 

This is only a WARN, so no one is forced to do this. But maybe this is a good best practice if you don't have any reason to do otherwise?

Here is Martian Mono, which is perfectly centered (Top: 42px, Bottom: 42px):

<img width="1687" alt="Screenshot 2023-08-13 at 10 07 07 PM" src="https://github.com/googlefonts/fontbakery/assets/5162664/ae38062f-def1-4238-9065-6bc84e947bc4">

Here is Shantell Sans, which is NOT perfectly centered, but within the threshold to pass this test (Top: 46px, Bottom: 46px):

<img width="1380" alt="Screenshot 2023-08-13 at 10 22 28 PM" src="https://github.com/googlefonts/fontbakery/assets/5162664/8de8e94b-fa63-4b35-b3af-12949b228274">

Here is Amiri, which is off by enough that it would get a `WARN`, note that this is visibly not centered (Top: 54px, Bottom: 62px):

<img width="1224" alt="Screenshot 2023-08-13 at 10 21 45 PM" src="https://github.com/googlefonts/fontbakery/assets/5162664/9976e29c-4e9b-4927-89e0-d1525eddb17b">

**Comment by tphinney** [comment]
Clarification: I was writing about Roman Shemin’s Twitter thread and _his version_ of the vertical metrics check, which is significantly different than what Rosa proposed. _Which one are you doing?_

**Comment by eliheuer** [comment]
I implemented the check almost exactly as described by Rosalie, including the error margin of 5%. 
```python

    upm = ttFont["head"].unitsPerEm
    error_margin = upm * 0.05
    cap_height = ttFont["OS/2"].sCapHeight
    ascender = ttFont["hhea"].ascent
    descender = ttFont["hhea"].descent
    top_margin = ascender - cap_height
    difference = abs(top_margin - abs(descender))

    vertically_centered = difference <= error_margin
```

**Comment by m4rc1e** [comment]
I don't like the use of `sCapHeight` since it's often set incorrectly. I don't mind this check being included but only as a warn. It's definitely something that should only apply to fonts intended for UI and only include scripts which have similar proportions (Amiri shows exactly why this doesn't work for most Arabic + Latin fonts).

**Comment by felipesanches** [comment]
I agree with @khaledhosny. To me this proposed check looks like trouble.

**Comment by felipesanches** [comment]
The in-person conversations at TypeCon 2023 convinced me it is a reasonable check.

**Comment by tphinney** [comment]
Yeah, as long as it is Rosalie’s version and not Roman Shemin’s it is… not crazy. 😛

**Comment by eliheuer** [comment]
This issue is a great example of this problem: https://github.com/mui/material-ui/issues/29965
